### PR TITLE
Tpetra: Fix #6455

### LINKED ISSUE #6455: Teko_testdriver_tpetra_MPI_1 failing in ATDM Power/GPU CUDA builds on 'ride' and 'waterman' starting 2019-12-13
CC: @trilinos/teko, @srajama1  (Trilinos Linear Solvers Product Lead), @bartlettroscoe, @fryeguy52, @brian-kelley

<Checklist>
<???: Add label "client: ATDM">
<???: Add label "ATDM Sev: Blocker" (by default but could be other "ATDM Sev: XXX")>
<???: Add label "type: bug"?>
<???: Add label for affected packages (e.g. "pkg: MueLu", "pkg: Tpetra", "pkg: Kokkos", etc.)>
<???: Add label "PA: ???Project Area???" (e.g. "PA: Linear Solvers", "PA: Data Services")>
<???: Add milestone "Initial cleanup of new ATDM ..." or "Keep promoted ATDM ...">
<???: Once GitHub Issue is created, add entries for tests to TrilinosATDMStatus/*.csv files>

## Next Action Status

<status-and-or-first-action>

## Description

As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-11-10&end=2019-12-16&filtercount=4&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-&field3=testname&compare3=61&value3=Teko_testdriver_tpetra_MPI_1&field4=status&compare4=61&value4=failed) the test:

* `Teko_testdriver_tpetra_MPI_1`

in the builds:

* `Trilinos-atdm-waterman-cuda-9.2-debug`
* `Trilinos-atdm-waterman-cuda-9.2-rdc-release-debug`
* `Trilinos-atdm-waterman-cuda-9.2-release-debug`
* `Trilinos-atdm-white-ride-cuda-10.1-gnu-7.2.0-release-debug`
* `Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-debug`
* `Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-rdc-release-debug`
* `Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-release-debug`

failing every day starting on testing day 2019-12-13.

As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2019-12-15&filtercount=4&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-&field3=buildname&compare3=63&value3=cuda&field4=testname&compare4=61&value4=Teko_testdriver_tpetra_MPI_1), this test only failes in builds that have `Trilinos_ENABLE_DEBUG=ON` set (i.e. with `debug` or `dbg` in the name) and does not fail in any full `opt`/`release` builds.

As shown in [this query](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&begin=2019-11-10&end=2019-12-16&filtercount=5&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-&field3=testname&compare3=61&value3=Teko_testdriver_tpetra_MPI_1&field4=status&compare4=61&value4=failed&field5=testoutput&compare5=97&value5=Tpetra%3A%3ACrsMatrix.*%3A%3AsetAllValues%3A%20myGraph_-.setAllIndices..%20threw%20an%20exception%3A%20CrsGraph%3A%3AsetAllIndices..%3A%20provided%20columnIndices%20are%20not%20sorted%20within%20rows%20on%20at%20least%20one%20process) (click "Show Matching Output" in upper right) all of these failing tests showed:

```
ERROR, rank 0, CrsGraph::setAllIndices(): provided columnIndices are not sorted!
terminate called after throwing an instance of 'std::runtime_error'
  what():  /home/jenkins/waterman/workspace/Trilinos-atdm-waterman-cuda-9.2-debug/SRC_AND_BUILD/Trilinos/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp:4117:

Throw number = 3

Throw test that evaluated to true: (true)

Tpetra::CrsMatrix<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Cuda, Kokkos::CudaUVMSpace> >::setAllValues: myGraph_->setAllIndices() threw an exception: CrsGraph::setAllIndices(): provided columnIndices are not sorted within rows on at least one process.
[waterman1:143826] *** Process received signal ***
[waterman1:143826] Signal: Aborted (6)
[waterman1:143826] Signal code:  (-6)
```

Looking at the new commits pulled that day, for example [here](https://testing-dev.sandia.gov/cdash/viewNotes.php?buildid=5084925#!#note3), it seems very likely that this failure was triggered by the changes in the commit:

```
a7eee78:  Tpetra: setAllIndices verifies sorting in debug
Author: Brian Kelley <bmkelle@sandia.gov>
Date:   Wed Dec 11 10:25:12 2019 -0700

M	packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp
M	packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp
D	packages/tpetra/core/test/MatrixMatrix/AddProfilingTest.cpp
M	packages/tpetra/core/test/MatrixMatrix/CMakeLists.txt
M	packages/tpetra/core/test/MatrixMatrix/MatrixMatrix_UnitTests.cpp
```

from the PR #6433 by @brian-kelley.

## Current Status on CDash

* [Status of `Teko_testdriver_tpetra_MPI_1` in all CUDA builds for current testing day](https://testing-dev.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercount=4&showfilters=1&filtercombine=and&field1=groupname&compare1=62&value1=Experimental&field2=buildname&compare2=65&value2=Trilinos-atdm-&field3=buildname&compare3=63&value3=cuda&field4=testname&compare4=61&value4=Teko_testdriver_tpetra_MPI_1)

## Steps to Reproduce

One should be able to reproduce this failure the machines 'white' (SON) 'ride' (SRN), or 'waterman' (SRC) as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

The exact commands to reproduce this failure on 'ride' or 'white', for example, should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh \
    Trilinos-atdm-white-ride-cuda-9.2-gnu-7.2.0-debug

$ cmake \
 -GNinja \
 -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
 -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Teko=ON \
 $TRILINOS_DIR

$ make NP=16

$ bsub -x -Is -q rhel7F -n 16 ctest -R Teko_testdriver_tpetra_MPI_1
```

### PR DESCRIPTION
**Pull request by brian-kelley**
@trilinos/tpetra 

## Motivation
For transposed operands, Tpetra::MatrixMatrix::add() didn't request the RowMatrixTransposer to sort rows because it can work with unsorted rows. But the transpose graph still reported ``isSorted() == true``. Semi-related to #6405 

## Testing

This was caught due to a test failure in Teko: #6455, though it actually has nothing to do with Teko or Cuda.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@brian-kelley wrote:
> For transposed operands, Tpetra::MatrixMatrix::add() didn't request the RowMatrixTransposer to sort rows because it can work with unsorted rows. But the transpose graph still reported `isSorted() == true`.

Wouldn't this be a bug in `RowMatrixTransposer`?

**Comment by brian-kelley** [comment]
@mhoemmen I guess so. The RowMatrixTransposer lets you ask to not sort indices, but then it calls a CrsGraph constructor that asks for (but doesn't verify) a sorted local graph.

**Comment by mhoemmen** [comment]
@brian-kelley wrote:
> I guess so. The RowMatrixTransposer lets you ask to not sort indices, but then it calls a CrsGraph constructor that asks for (but doesn't verify) a sorted local graph.

There's some reason relating to MueLu setup why the RowMatrixTransposer lets you ask not to sort indices.  Should we pass that reason along to CrsGraph's constructor?

**Comment by mhoemmen** [comment]
@brian-kelley If I remember right, it was @csiefer2 who added that option to RowMatrixTransposer as an optimization, because the matrix in question is some intermediate term that doesn't need sorted rows.

**Comment by brian-kelley** [comment]
@mhoemmen Yeah, if I could add a "bool sorted" parameter to that CrsGraph constructor, that would  completely solve the problem. It could default to true. If we're doing that, it should also be added to the constructors that take individual CRS views and then call setAllIndices().

@bartlettroscoe I agree, I will add a test tomorrow.

**Comment by mhoemmen** [comment]
@brian-kelley wrote:
> Yeah, if I could add a "bool sorted" parameter to that CrsGraph constructor, that would completely solve the problem.

CrsGraph does have constructors that take a ParameterList; that might be a place to add it.  It should default to `true`, since `add` is a public interface (apparently -- Teko is calling it) and thus should return a sensible CrsGraph by default.

**Comment by brian-kelley** [comment]
@mhoemmen By "sensible", do you mean that the CrsGraph ctor should call sortCrsEntries if the parameter list says the input's not sorted?

**Comment by mhoemmen** [comment]
@brian-kelley "Sensible" means that the result of `add` should be sorted by default, while still preserving the option that MueLu wants of having a nonsorted CrsGraph.  I don't like generally how MueLu interacts with Tpetra objects, but that's too big of an issue.

**Comment by mhoemmen** [comment]
@bartlettroscoe I appreciate the commitment to testing, but this change actually makes testing easier by removing the need for test cases.  The only reason I haven't approved yet is because I want to hear from @trilinos/muelu about whether this will affect performance a lot.

**Comment by brian-kelley** [comment]
@mhoemmen After I add support for "sorted" in the ParameterList, I still want to add a new test for add() in the transposed A/B operand cases, because this should have been caught by MatrixMatrix tests.

**Comment by brian-kelley** [comment]
The result of add() having sorted rows was always intended to be an invariant (except when we realized it was broken because of the setAllValues/fillComplete logic, in #6405).

**Comment by mhoemmen** [comment]
@brian-kelley Looks like MueLu is OK with this, so I approved.  @bartlettroscoe , sorry to dismiss your review -- I think this change actually reduces the need for tests, by eliminating a possible state of CrsGraph.

**Comment by brian-kelley** [comment]
@mhoemmen This PR does fix the original issue, but if you'd like I could still add this change to the Graph/Matrix constructors that take local CRS + maps + paramlist:

```
if(!params->get("sorted"))
    isSorted_ = false;
else
    isSorted_ = true;
```

This is probably still worth doing since the RowMatrixTransposer still produces graphs that "lie" about whether they're sorted.

**Comment by brian-kelley** [comment]
Also, it looks like the only failure in PR testing here was due to #6361, which is not related to this commit.

**Comment by bartlettroscoe** [comment]
CC: @trilinos/framework 

> Also, it looks like the only failure in PR testing here was due to #6361, which is not related to this commit.

@brian-kelley, I fear that until someone addresses #6361, none of these PRs that trigger the testing of MueLu will be able to be merged.

**Comment by brian-kelley** [comment]
@mhoemmen Actually, I need to do one more thing, you don't need to review yet.

File: packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp [hunk 1]
```diff
@@ -271,12 +273,18 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList>& params)
   RCP<const export_type> myExport = origImport.is_null () ?
     Teuchos::null : rcp (new export_type (*origImport));
 
+  RCP<Teuchos::ParameterList> graphParams = Teuchos::null;
+  if(!sort) {
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp]: I think this is OK.

File: packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp [hunk 2]
```diff
@@ -271,12 +273,18 @@ createTransposeLocal (const Teuchos::RCP<Teuchos::ParameterList>& params)
   RCP<const export_type> myExport = origImport.is_null () ?
     Teuchos::null : rcp (new export_type (*origImport));
 
+  RCP<Teuchos::ParameterList> graphParams = Teuchos::null;
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_RowMatrixTransposer_def.hpp]: `RCP`'s default value is null, so you don't need the assignment here.

**Comment by brian-kelley** [comment]
@bartlettroscoe One new Tpetra test ("locally_unsorted") for MatrixMatrix::add() now exercises the case where the inputs are unsorted and have the same column map. There was never a problem with that kernel, but it needed test coverage. 

I can confirm that the other new test ("tranposed_b") would have caught the original bug, where the graph lied about being sorted.

File: packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp [hunk 3]
```diff
@@ -749,13 +749,16 @@ namespace Tpetra {
...
     , lowerTriangular_ (false)
     , upperTriangular_ (false)
-    , indicesAreSorted_(true)
     , noRedundancies_(true)
     , haveLocalConstants_ (false)
     , haveGlobalConstants_ (false)
     , sortGhostsAssociatedWithEachProcessor_(true)
   {
     staticAssertions ();
+    if(!params.is_null() && !params->get<bool>("sorted"))
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp]: If `params` is not null but "sorted" isn't in it, then one-argument `get` will throw.  (Two-argument `get` sets a default value if the parameter doesn't exist.)  You could try this instead:
```c++
if (! params.is_null () && params->isParameter("sorted") &&
  ! params->get<bool>("sorted")) // etc.
```

File: packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp [hunk 4]
```diff
@@ -784,13 +787,16 @@ namespace Tpetra {
...
     , lowerTriangular_ (false)
     , upperTriangular_ (false)
-    , indicesAreSorted_ (true)
     , noRedundancies_ (true)
     , haveLocalConstants_ (false)
     , haveGlobalConstants_ (false)
     , sortGhostsAssociatedWithEachProcessor_ (true)
   {
     staticAssertions ();
+    if(!params.is_null() && !params->get<bool>("sorted"))
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp]: See above.

File: packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp [hunk 5]
```diff
@@ -936,6 +946,12 @@ namespace Tpetra {
 
     k_lclInds1D_ = lclGraph_.entries;
     k_rowPtrs_ = lclGraph_.row_map;
+
+    if(!params.is_null() && !params->get<bool>("sorted"))
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp]: See above.

**Comment by brian-kelley** [comment]
@bartlettroscoe It looks a random failure in MueLu_ParameterListInterpreter (on intel 17, but not the other builds) was the only issue with the last PR test. If that doesn't get resolved soon could we just manually merge this PR? It fixes a bug and a test failure.

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe It looks a random failure in MueLu_ParameterListInterpreter (on intel 17, but not the other builds) was the only issue with the last PR test. If that doesn't get resolved soon could we just manually merge this PR? It fixes a bug and a test failure.

@brian-kelley, I don't have the permissions to do that.  I am no different than any other Trilinos developer.  Until someone resolves this failing test as part of #6361 (and reverting the PR #6432 would likely do this), I don't think anyone's PR is going to get merged.  You will need to take this up with @trilinos/framework.

**Pull request merged**